### ISSUE TITLE
[API Proposal]: `Obsolete JsonNode.WriteTo

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#108643]
### Background and motivation

`JsonNode` is a meant to be a higher-level API. It can parse from `Stream` but it can't write back to one. Instead, users need to create a Utf8JsonWriter themselves, which isn't very consistent. Most of our APIs that deal with data offer `Stream` as the canonical exchange type.

### API Proposal

```C#
namespace System.Text.Json.Nodes;

public abstract class JsonNode
{
    public void WriteTo(Stream stream, JsonSerializerOptions? options = null);
}
```

### API Usage

```csharp
JsonNode node = GetNode();
using Stream stream = File.Create(somePath);
node.WriteTo(stream);
```

**Comment by hez2010** [comment]
Should a `WriteToAsync` also be added for asynchronous operation?

**Comment by eiriktsarpalis** [comment]
`Utf8JsonWriter` supports wrapping `Stream` so this is already possible today:

```C#
JsonNode node = JsonNode.Parse("{}")!;
MemoryStream stream = new();
Utf8JsonWriter writer = new(stream);
node.WriteTo(writer);
```

**Comment by bartonjs** [comment]
* While it's hard to discover, this proposal is entirely redundant with JsonSerializer.Serialize{Async}(stream, node, options?)
* We should instead mark the existing WriteTo method as Obsolete, directing potential callers to JsonSerializer.

```C#
namespace System.Text.Json.Nodes;

public abstract class JsonNode
{
    [Obsolete("Something about use JsonSerializer.Serialize{Async} instead.", DiagnosticId = NextUp)]
    public void WriteTo(Utf8JsonWriter stream, JsonSerializerOptions? options = null);
}
```

**Comment by campersau** [comment]
There is currently no `JsonSerializer.SerializeAsync` version which takes an `Utf8JsonWriter`. The current `JsonSerializer.Serialize` does a synchronous `Flush` while `JsonNode.WriteTo` does not.
https://github.com/dotnet/runtime/blob/abf8e94296b7842e4d8ba7c9791eaffaf63d2a09/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs#L37
When I switched to `JsonSerializer.Serialize` in my aspnet core project it throwed a runtime exception because aspnet core does not allow synchronous IO by default https://learn.microsoft.com/aspnet/core/fundamentals/servers/kestrel/options

Would it be possible to also add these APIs which will use `FlushAsync`?

```c#
public static void SerializeAsync<TValue>(Utf8JsonWriter writer, TValue value, JsonTypeInfo<TValue> jsonTypeInfo, CancellationToken cancellationToken = default);
public static void SerializeAsync<TValue>(Utf8JsonWriter writer, TValue value, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default);
```

**Comment by eiriktsarpalis** [comment]
> Would it be possible to also add these APIs which will use `FlushAsync`?

It wouldn't address the underlying problem you are experience. Namely, `Utf8JsonWriter` (and consequently all the converter implementations that consume it) do not expose async write methods so it necessarily must exercise the sync methods of the underlying stream. Async serialization in STJ is implemented differently, bypassing the `Stream`-based APIs in Utf8JsonWriter altogether.

**Comment by eiriktsarpalis** [comment]
I don't believe it is appropriate for us to obsolete this method. It is the core virtual method that we use to define serialization semantics for all the types deriving from `JsonNode`. In hindsight, this should have been marked `protected` but obsoletion just doesn't feel right to me.

**Comment by eiriktsarpalis** [comment]
Closing per decision reached in https://github.com/dotnet/runtime/pull/111392#discussion_r2137366957
